crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -356,22 +359,136 @@ impl PublishFlakeRef {
 
         Ok(publish_flake_ref)
     }
-}
 
-impl TryFrom<FlakeRef> for PublishFlakeRef {
-    type Error = ConvertFlakeRefError;
+    /// Resolve a git+file flake ref to a git+https reference
+    ///
+    /// Reproducing a snapshot from source requires access to the original repo.
+    /// Including a local file reference in the snapshot,
+    /// means its practically only possible to reproduce for the original creator.
+    /// Thus, we resolve a local branch to it's upstream remote and branch.
+    ///
+    /// For local repositories we also check
+    /// whether the repository contains any uncommitted changes
+    /// and is in sync with its upstream branch.
+    async fn from_git_file_flake_ref(
+        file_ref: GitRef<protocol::File>,
+        nix: &NixCommandLine,
+    ) -> Result<Self, ConvertFlakeRefError> {
+        // Get nix metadata for the referred flake
+        // Successfully acquiring metadata proves the path
+        // - is in fact a flake
+        // - is a git repository
+        // - the path is not dirty (no uncommitted changes)
+        let local_metadata = {
+            let command = runix::command::FlakeMetadata {
+                flake_ref: Some(FlakeRef::GitPath(file_ref.clone()).into()),
+                ..Default::default()
+            };
+            command
+                .run_typed(nix, &Default::default())
+                .await
+                .map_err(ConvertFlakeRefError::LocalFlakeMetadata)?
+        };
+
+        if local_metadata.revision.is_none() {
+            Err(ConvertFlakeRefError::LocalFlakeDirty)?;
+        }
+
+        // Create a handle to the git repo
+        let repo = Git::discover(file_ref.url.path())
+            .await
+            .map_err(|_| ConvertFlakeRefError::RepoNotFound(file_ref.url.path().into()))?;
+
+        // Get the upstream branch information.
+        // This is essentialy
+        //
+        //   upstream_ref = git rev-parse @{u}
+        //   (remote_name, branch_name) = split_once "/" upstream_ref
+        //   upstream_url = git remote get-url ${remote_name}
+        //   upstream_rev = git ls-remote ${remote_name} ${branch_name}
+        //
+        // The current branch MUST have an upstream ref configured
+        // which resolves to a valid rev on the remote.
+        let (remote_name, remote_url, remote_branch, remote_revision) = repo
+            .get_origin()
+            .await
+            .map_err(ConvertFlakeRefError::NoRemote)?;
+
+        // Ensure the remote branch exists
+        let remote_revision = remote_revision.ok_or(ConvertFlakeRefError::RemoteBranchNotFound)?;
+
+        info!("Resolved local flake to remote '{remote_name}:{remote_branch}' at '{remote_url}'");
```

Comment:
Do we display INFO level logs to users? If so, this should probably be a DEBUG level log.

Reply:
Yeah at this level debug is probably better.
There were some issue with filtering lib related logs anyway (i dont know if those got resolved lately)